src/lib/core/projection/projection.ts

```diff
@@ -0,0 +1,56 @@
+import {Injectable, Directive, ModuleWithProviders, NgModule, ElementRef} from '@angular/core';
+
+
+@Directive({
+  selector: 'md-host'
+})
+export class MdProjectionHostDirective {
+  constructor(public ref: ElementRef) {}
+}
+
+
+@Injectable()
+export class MdProjectionService {
+  private _replaceWith(toReplaceEl: HTMLElement, otherEl: HTMLElement) {
+    toReplaceEl.parentElement.replaceChild(otherEl, toReplaceEl);
+  }
+
+  /**
+   * Project an element into a host element.
+   */
+  project(ref: ElementRef, host: MdProjectionHostDirective): void {
+    const componentEl = ref.nativeElement;
+    const hostEl = host.ref.nativeElement;
+    const childNodes = componentEl.childNodes;
+    let child = childNodes[0];
+
+    // Replace ComponentEl by all its children, in place.
+    this._replaceWith(componentEl, child);
+    while (childNodes.length) {
```

Comment:
Reading `length` on a `NodeList` is notoriously slow; it's better to cache the length and `for` loop.
(here and below)